[TN] Error Status Code 200 (#763)
apollographql/docs: src/content/technotes/TN0047-error-status-code-200.mdx

---
title: Error! Status Code 200?
subtitle: Granular error handling in your API
description: Learn GraphQL uniquly handles errors and how that compares to traditional REST APIs.
published: 2024-02-09
id: TN0047
tags: [best-practices, error-handling]
---

In the realm of API development, HTTP status codes serve as crucial indicators of request success or failure.
However, when it comes to GraphQL, the traditional approach of using REST/HTTP error codes becomes inadequate.
GraphQL introduces a paradigm shift in how APIs are designed and operated, rendering conventional status codes insufficient for capturing the nuances of GraphQL operations, especially in the context of parallel operations.

>

> > TLDR; The GraphQL 200 status codes signify the successful processing of a request at the Apollo Router level, but they do not encapsulate the outcome of individual operations or the presence of errors within a GraphQL response.

Reviewer: ```suggestion
> **TLDR;** The GraphQL 200 status codes signify the successful processing of a request at the Apollo Router level, but they do not encapsulate the outcome of individual operations or the presence of errors within a GraphQL response.
```

> > TLDR; The GraphQL 200 status codes signify the successful processing of a request at the Apollo Router level, but they do not encapsulate the outcome of individual operations or the presence of errors within a GraphQL response.

Reviewer: ```suggestion
> **TLDR;** The GraphQL 200 status codes signify the client has sent a request, the operation is valid GraphQL against the schema, and the GraphQL server was able to return a spec-compliant GraphQL response, but they do not encapsulate the outcome of individual fields or the presence of errors within the GraphQL response.

```

> Attempting to shoehorn HTTP error codes into the GraphQL paradigm overlooks the unique characteristics and capabilities of GraphQL, such as granular error handling, parallel execution, and asynchronous patterns.
> 
> You can learn more about status codes in GraphQL by reading the [GraphQL-over-HTTP](https://graphql.github.io/graphql-over-http/draft/) specification.
>

## Nature of GraphQL Operations

In a typical REST API, each endpoint corresponds to a specific resource or action, making it straightforward to map HTTP status codes to the outcome of each request.
However, GraphQL operates differently. Instead of multiple endpoints, GraphQL has a single endpoint that serves as an entry point for executing queries, mutations, and subscriptions.
This unified approach allows clients to request precisely the data they need in a single round trip to the server.

## Granular Error Handling

GraphQL offers granular error handling through its response structure.

> In a GraphQL response, along with the data requested by the client, the server can include an array of errors, each containing detailed information about what went wrong.

Reviewer: ```suggestion
In a GraphQL response,  you have two key fields: the `data` field represents all the information the server could respond with that matches the clients request, and `errors` field is an array containing detailed information about what may have gone wrong when making the request. If the server is able to return a response in this shape, it is matching that GraphQL spec, even if it only contains the `errors` field.
```

---
title: Error! Status Code 200?
subtitle: Granular error handling in your API

> description: Learn GraphQL uniquly handles errors and how that compares to traditional REST APIs.

Reviewer 2: ```suggestion
description: Learn how GraphQL uniquely handles errors and how that compares to traditional REST APIs.
```

published: 2024-02-09
id: TN0047
tags: [best-practices, error-handling]
---

In the realm of API development, HTTP status codes serve as crucial indicators of request success or failure.
However, when it comes to GraphQL, the traditional approach of using REST/HTTP error codes becomes inadequate.
GraphQL introduces a paradigm shift in how APIs are designed and operated, rendering conventional status codes insufficient for capturing the nuances of GraphQL operations, especially in the context of parallel operations.

>

> > **TLDR;** The GraphQL 200 status codes signify the client has sent a request, the operation is valid GraphQL against the schema, and the GraphQL server was able to return a spec-compliant GraphQL response, but they do not encapsulate the outcome of individual fields or the presence of errors within the GraphQL response.

Reviewer 2: You have a great succinct summary that works well here.

```suggestion
> **TLDR;** In GraphQL, the presence of a 200 status code indicates that the request was successful from the Apollo Router's perspective, but it doesn't necessarily mean that the client received all the data it requested without any errors.

The GraphQL 200 status codes don't encapsulate the outcome of individual fields or the presence of errors within the GraphQL response. Rather, they signify that the client has sent a request, the operation is valid GraphQL against the schema, and the GraphQL server was able to return a spec-compliant GraphQL response. 
```

>

> > Attempting to shoehorn HTTP error codes into the GraphQL paradigm overlooks the unique characteristics and capabilities of GraphQL, such as granular error handling, parallel execution, and asynchronous patterns.

Reviewer 2: ```suggestion
> Shoehorning HTTP error codes into the GraphQL paradigm overlooks the unique characteristics and capabilities of GraphQL, such as granular error handling, parallel execution, and asynchronous patterns.
```

> 

> > You can learn more about status codes in GraphQL by reading the [GraphQL-over-HTTP](https://graphql.github.io/graphql-over-http/draft/) specification.

Reviewer 2: ```suggestion
> To learn more about GraphQL status codes, see the [GraphQL-over-HTTP](https://graphql.github.io/graphql-over-http/draft/) specification.
```

>

> ## Nature of GraphQL Operations

Reviewer 2: ```suggestion
## Nature of GraphQL operations
```

In a typical REST API, each endpoint corresponds to a specific resource or action, making it straightforward to map HTTP status codes to the outcome of each request.
However, GraphQL operates differently. Instead of multiple endpoints, GraphQL has a single endpoint that serves as an entry point for executing many different queries, mutations, and subscriptions. You can even execute more then one query in a single request.
This unified approach allows clients to request precisely the data they need in a single round trip to the server.

## Granular Error Handling

GraphQL offers granular error handling through its response structure.

> In a GraphQL response,  you have two key fields: the `data` field represents all the information the server could respond with that matches the clients request, and `errors` field is an array containing detailed information about what may have gone wrong when making the request. If the server is able to return a response in this shape, it is matching that GraphQL spec, even if it only contains the `errors` field.

Reviewer 2: ```suggestion
A GraphQL response contains two key fields: 
- The `data` field represents all the information the server could respond with that matches the clients request.
- The `errors` field is an array containing detailed information about what may have gone wrong when making the request. 

A server returns a response in this shape, even if it only contains the `errors` field.
```

This level of granularity enables clients to understand and react to errors more effectively than traditional HTTP status codes would allow.

> ## Parallel Execution and Partial Results

Reviewer 2: ```suggestion
## Parallel execution and partial results
```

> One of the most powerful features of GraphQL is its ability to execute multiple operations in parallel.
> This means that even if one part of a GraphQL operation fails, other parts may still succeed.

Reviewer 2: ```suggestion
One of the most powerful features of GraphQL is its ability to execute multiple operations in parallel: if one part of a GraphQL operation fails, other parts may still succeed.
```

In contrast, REST APIs typically follow a serial execution model, where one failed request can halt the entire process.

Therefore, using HTTP status codes to convey the outcome of parallel GraphQL operations would be impractical and misleading.

> 

Reviewer: I think we still need a section calling out when non-200s are used. The GraphQL over http spec allows it and so we don't want some one to take away from the this note that you ALWAYS have to return 200

## Summary

In GraphQL, the presence of a 200 status code indicates that the request was successful from the Apollo Router's perspective.

> This does not necessarily mean that the client received all the data it requested without any errors.

Reviewer 2: ```suggestion
This doesn't necessarily mean that the client received all the data it requested without any errors.
```
